napari/layers/_points_layer/model.py

```diff
@@ -375,37 +555,133 @@ def _select_point(self, indices):
     def _set_view_slice(self):
         """Sets the view given the indices to slice with."""
 
-        in_slice_points, matches, scale = self._slice_points(self.indices)
+        in_slice_points, indices, scale = self._slice_points(self.indices)
 
         # Display points if there are any in this slice
         if len(in_slice_points) > 0:
             # Get the point sizes
-            sizes = (self._size[matches, -2:].mean(axis=1) * scale)[::-1]
+            sizes = self.size_array[indices, -2:].mean(axis=1) * scale
 
             # Update the points node
-            data = np.array(in_slice_points)[::-1] + 0.5
+            data = np.array(in_slice_points) + 0.5
 
         else:
             # if no points in this slice send dummy data
             data = np.empty((0, 2))
-            sizes = 0
+            sizes = [0]
         self._points_view = data
         self._sizes_view = sizes
+        self._indices_view = indices
+        # Make sure if changing planes any selected points not in the current
+        # plane are removed
+        selected = []
+        for c in self._cur_selected:
+            if c in self._indices_view:
+                ind = list(self._indices_view).index(c)
+                selected.append(ind)
+        self._selected_points = selected
+        self._selected_box = self.interaction_box(self.selected_points)
+        if len(self.selected_points) > 0:
+            self._cur_selected = self._indices_view[self.selected_points]
+        else:
+            self._cur_selected = []
+
+        if len(data) > 0:
+            edge_color = [
+                self._edge_color_list[i] for i in self._indices_view[::-1]
+            ]
+            face_color = [
+                self._face_color_list[i] for i in self._indices_view[::-1]
+            ]
+        else:
+            edge_color = 'white'
+            face_color = 'white'
 
-        self._node.set_data(
-            data[:, [1, 0]],
-            size=sizes,
+        self._node._subvisuals[2].set_data(
+            data[::-1, [1, 0]],
+            size=sizes[::-1],
             edge_width=self.edge_width,
             symbol=self.symbol,
-            edge_color=self.edge_color,
-            face_color=self.face_color,
+            edge_color=edge_color,
+            face_color=face_color,
             scaling=True,
         )
         self._need_visual_update = True
+        self._set_highlight(force=True)
         self._update()
-        self.status = self.get_message(self.coordinates, self._selected_points)
+        self.status = self.get_message(self.coordinates, self._hover_point)
         self._update_thumbnail()
 
+    def _set_highlight(self, force=False):
```

Comment:
agreed - we havn't seperated out the vispy views yet from the individual layers. It's the last remaining thing to do in our MVC refactor and we can start getting to it soon after this PR merges. At that point this will probably go along into the view file